Add tests for org.zaproxy.zap.utils.ByteBuilder

Braavos96 opened the pull request:
I've analysed your codebase and noticed that `org.zaproxy.zap.utils.ByteBuilder` is not fully tested.
I've written some tests for the methods in this class with the help of [Diffblue Cover](https://www.diffblue.com/opensource).

Hopefully, these tests will help you detect any regressions caused by future code changes. If you would find it useful to have additional tests written for this repository, I would be more than happy to look at other classes that you consider important.

psiinon commented:
@Braavos96 thank you! Yes, we definitely would like more tests like this :D

psiinon commented:
@Braavos96 its failing this build, but just due to formatting issues:
Run 'gradlew spotlessApply' to fix these violations :)

kingthorin commented:
You should also refer to: https://github.com/zaproxy/zaproxy/wiki/DevGuidelines#unit-test-specific-guidelines (Which basically talks about test method names that describe the individual tests as well as comments to break things down `//Given`, `// When`, `//Then`

Thanks for the contribution :+1:

thc202 commented on zap/src/test/java/org/zaproxy/zap/utils/ByteBuilderTest.java:
```diff
@@ -0,0 +1,241 @@
+    @Test
+    public void testEnsureCapacity() {
+        ByteBuilder byteBuilder = new ByteBuilder(5);
+        Assert.assertEquals(5, byteBuilder.capacity());
```
This is already being tested, please, add the (new) tests to existing class.

kingthorin replied:
`ByteBuilderUnitTest.java`

Braavos96 replied:
Moved test to `ByteBuilderUnitTest.java`

Braavos96 commented:
@psiinon Thank you for the tip. It seemed to have built after running `gradlew spotlessApply`. 
No problem :) do you have any specific class in mind?

@kingthorin I had a look at the wiki I think I was able to adhere to those standards. However, I did something different in the method `shouldIncreaseCapacityFrom5To20` as I put 2 `asserts` in there

@thc202 I have moved the tests to `ByteBuilderUnitTest.java` and removed the original test file `ByteBuilderTest.java`

kingthorin commented:
@Braavos96 thanks. For that one weird case you mentioned the first assertion can probably be removed since the "test" overlaps with `shouldBeInitializedWithGivenCapacity`. So for my 2 cents something like the following probably does what's needed (though I'm open to discussion/debate):

```java
    @Test
    public void shouldIncreaseCapacityFrom5To20() {
        // given
        byteBuilder = new ByteBuilder(5);
        // when
        byteBuilder.ensureCapacity(20);
        capacity = byteBuilder.capacity();
        // then
        assertThat(capacity, is(20));
    }
```

Braavos96 commented:
@kingthorin I have implemented your changes :)

thc202 commented on zap/src/test/java/org/zaproxy/zap/utils/ByteBuilderUnitTest.java:
```diff
@@ -65,4 +69,314 @@ public void shouldBeInitializedWithDoubleCapacityOfGivenArray() {
     }
+
+    @Test
+    public void shouldIncreaseCapacityFrom5To20() {
```
It should be e.g. `shouldIncreaseCapacityWhenBiggerThanActual`, which values are used it's not important. (Ideally it should have other tests with negative and lower than actual capacity.)

Braavos96 replied:
Test name changed

thc202 commented on zap/src/test/java/org/zaproxy/zap/utils/ByteBuilderUnitTest.java:
```diff
@@ -65,4 +69,314 @@ public void shouldBeInitializedWithDoubleCapacityOfGivenArray() {
+    }
+
+    @Test
+    public void shouldHaveASizeOf0() {
```
It would be better with e.g. `shouldHaveZeroSizeByDefault`, otherwise it's not clear what/when behaviour is expected, specially with the following method (i.e. should have a size of 0 or 3?).

Braavos96 replied:
Test name changed

thc202 commented on zap/src/test/java/org/zaproxy/zap/utils/ByteBuilderUnitTest.java:
```diff
@@ -65,4 +69,314 @@ public void shouldBeInitializedWithDoubleCapacityOfGivenArray() {
+    }
+
+    @Test
+    public void shouldHaveASizeOf3() {
```
e.g. `shouldHaveSizeOfGivenArray`

Braavos96 replied:
Test name changed

thc202 commented on zap/src/test/java/org/zaproxy/zap/utils/ByteBuilderUnitTest.java:
```diff
@@ -65,4 +69,314 @@ public void shouldBeInitializedWithDoubleCapacityOfGivenArray() {
+    }
+
+    @Test
+    public void shouldHaveALengthOf3() {
```
Same here (and following), it's not clear why it should have a length of 3 (nor the what, it's asserting the length of the string converted to bytes).

Braavos96 replied:
Test name changed

thc202 commented:
The commit does not look right (it's amending a PR merge?).

kingthorin commented:
@Braavos96 

> The commit does not look right (it's amending a PR merge?).

Why is this building on PR (feature branches) not `develop`?

Braavos96 commented:
> @Braavos96
> 
> > The commit does not look right (it's amending a PR merge?).
> 
> Why is this building on PR (feature branches) not `develop`?

I am sorry but I don't understand what you mean by this comment?
Is it because I used a clone of the develop branch instead of the master?

thc202 commented:
No, if you look at the commit it's amending a merge commit from a PR (the commit message mentions the merge and the author of the commit is @psiinon instead of yourself).

Braavos96 commented:
> No, if you look at the commit it's amending a merge commit from a PR (the commit message mentions the merge and the author of the commit is @psiinon instead of yourself).

Hi, Apologies. but the DCO requires the commit to be signed off and as soon as I do that the Author changes. I am not sure what I should do about this.

kingthorin commented:
@Braavos96 are you using eclipse to do the commit/push ?

kingthorin commented:
If so don't click the first button (with the `A` that's amend commit), click the second one with the pencil/pen icon (that's "Add Signed-off By").

Either way what you need to add is a line in the commit message like:
```text
Add unit tests for org.zaproxy.zap.utils.ByteBuilder

These tests were written using Diffblue Cover.

Signed-off-by: Braavos96 <Braavos96@users.noreply.github.com>
```

Braavos96 commented:
> @Braavos96 are you using eclipse to do the commit/push ?

Nope.. I am using the ubuntu CLI to do the commit. I was using the commands
`git commit
git push -f origin add-diffblue-tests`

I will add the line to the new commit

kingthorin commented:
In the CLI apparently it's `--signoff`

https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

Braavos96 commented:
> In the CLI apparently it's `--signoff`
> 
> https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

I was first using that command but the DCO seems to have passed now

kingthorin commented:
Ya I set it to pass. If you have other commits (after second review), then you can use `--sign`. (I was trying to find docs on it quickly but couldn't, it might require user.name and user.email to be set in your git config.)

Braavos96 commented:
> Ya I set it to pass. If you have other commits (after second review), then you can use `--sign`. (I was trying to find docs on it quickly but couldn't, it might require user.name and user.email to be set in your git config.)

ohh ok thank you :)

Braavos96 commented:
Hi, just following up and wondering if you would like me to make any changes to the PR?

thc202 commented:
There are still some tests that are not clear, e.g.`shouldThrowAnArrayIndexOutOfBoundsExceptionForEmptyByteArray`, `shouldThrowAnArrayIndexOutOfBoundsExceptionForEmptyCharArray`, `shouldAppendByteBufferToByteBuilder`.
(The suffix `ToByteBuilder` in the test names could be removed, that's implied by the test class.)

For the record, the `toString()` is not correct, it should not use the whole `array` (only to the `size`).

Braavos96 commented:
> There are still some tests that are not clear, e.g.`shouldThrowAnArrayIndexOutOfBoundsExceptionForEmptyByteArray`, `shouldThrowAnArrayIndexOutOfBoundsExceptionForEmptyCharArray`, `shouldAppendByteBufferToByteBuilder`.
> (The suffix `ToByteBuilder` in the test names could be removed, that's implied by the test class.)
> 
> For the record, the `toString()` is not correct, it should not use the whole `array` (only to the `size`).

Hi,
Sorry i don't understand the first and last point you made.
Do you mean the test method names are not clear or the tests itself?

What do you mean `only to the size`?

The suffix has been removed.

thc202 commented:
The names, the first two it's not clear that the exception is thrown when appending specific values of the empty array, the last says `ByteBuffer` but should be `ByteBuilder`. 

That `toString()` returns `new String(array, 0 , size)`.

Braavos96 commented:
> The names, the first two it's not clear that the exception is thrown when appending specific values of the empty array, the last says `ByteBuffer` but should be `ByteBuilder`.
> 
> That `toString()` returns `new String(array, 0 , size)`.

Hi,
I changed the method names
Also for `toString()` I used `substring()` to only achieve the return value `new String(array, 0 , size)`

thc202 commented:
The method `shouldAppendByteBuffer` was not renamed.

Sorry, I didn't mean you needed to change the test (though it could just compare the string value), the actual `ByteBuilder.toString()` is what needs to be changed/fixed. I just mentioned that for the record/us (the FAQ entry "Will the tests find bugs in my code?" says "which has led to the repo owner fixing that bug.", which I read as "it's up to us to fix it").

thc202 commented:
Thanks!
